lib/evmone/instructions_calls.cpp

```diff
@@ -226,6 +224,43 @@ Result create3(StackTop stack, int64_t gas_left, ExecutionState& state, code_ite
     if (!check_memory(gas_left, state.memory, input_offset_u256, input_size_u256))
         return {EVMC_OUT_OF_GAS, gas_left};
 
+    bytes_view initcontainer;
+    if constexpr (Op == OP_CREATE3)
+    {
+        const auto initcontainer_index = uint8_t{pos[1]};
+        pos += 2;
+        const auto& container = state.original_code;
+        const auto eof_header = read_valid_eof1_header(state.original_code);
+        initcontainer = eof_header.get_container(container, initcontainer_index);
+    }
+    else
+    {
+        pos += 1;
+
+        const auto initcode =
+            state.get_tx_initcode_by_hash(intx::be::store<evmc::bytes32>(initcode_hash));
+        if (initcode.data == nullptr)
+            return {EVMC_SUCCESS, gas_left};  // "Light" failure
+
+        initcontainer = {initcode.data, initcode.size};
+
+        // Charge for initcode validation.
+        constexpr auto initcode_word_cost_validation = 2;
+        const auto initcode_cost_validation =
+            num_words(initcontainer.size()) * initcode_word_cost_validation;
+        if ((gas_left -= initcode_cost_validation) < 0)
+            return {EVMC_OUT_OF_GAS, gas_left};
+
+        const auto error_subcont = validate_eof(state.rev, initcontainer);
```

Comment:
It might make sense to move this validation after endowment check below, but not sure.

Reply:
Note that moving it would mean in case of invalid container, we still pay for hashing that never happens.
(But it is also true for balance check even in CREATE2)

Reply:
Would make sense to move gas spending for hash before validation. My thinking is if the contract can't be deployed bcs of gas I would like to know that before starting validation.

Reply:
I don't have a strong opinion on this. Charging gas on each logical step performed seemed like commons sense, but I'm inclined to agree that the consistency reasons should take priority over this.

Reply:
I moved validation to later and made a change to the spec: https://github.com/ipsilon/eof/pull/66